src/IO/WriteBufferFromS3.cpp

```diff
@@ -72,120 +191,264 @@ WriteBufferFromS3::WriteBufferFromS3(
     const String & key_,
     const S3Settings::RequestSettings & request_settings_,
     std::optional<std::map<String, String>> object_metadata_,
-    size_t buffer_size_,
     ThreadPoolCallbackRunner<void> schedule_,
     const WriteSettings & write_settings_)
-    : BufferWithOwnMemory<WriteBuffer>(buffer_size_, nullptr, 0)
-    , bucket(bucket_)
+    : bucket(bucket_)
     , key(key_)
     , request_settings(request_settings_)
     , upload_settings(request_settings.getUploadSettings())
+    , write_settings(write_settings_)
     , client_ptr(std::move(client_ptr_))
     , object_metadata(std::move(object_metadata_))
-    , upload_part_size(upload_settings.min_upload_part_size)
+    , buffer_allocation_policy(request_settings_.getUploadSettings())
     , schedule(std::move(schedule_))
-    , write_settings(write_settings_)
 {
+    LOG_TRACE(log, "Create WriteBufferFromS3, {}", getLogDetails());
+
     allocateBuffer();
 }
 
 void WriteBufferFromS3::nextImpl()
 {
-    if (!offset())
-        return;
+    LOG_TRACE(log, "nextImpl with incoming data size {}, memory buffer size {}. {}", offset(), memory.size(), getLogDetails());
 
-    /// Buffer in a bad state after exception
-    if (temporary_buffer->tellp() == -1)
-        allocateBuffer();
+    if (is_prefinalized)
+        throw Exception(
+                ErrorCodes::LOGICAL_ERROR,
+                "Cannot write to prefinalized buffer for S3, the file could have been created with PutObjectRequest");
 
-    size_t size = offset();
-    temporary_buffer->write(working_buffer.begin(), size);
+    hidePartialData();
 
-    ProfileEvents::increment(ProfileEvents::WriteBufferFromS3Bytes, offset());
-    last_part_size += offset();
-    if (write_settings.remote_throttler)
-        write_settings.remote_throttler->add(offset(), ProfileEvents::RemoteWriteThrottlerBytes, ProfileEvents::RemoteWriteThrottlerSleepMicroseconds);
+    if (available() > 0)
+        return;
 
-    /// Data size exceeds singlepart upload threshold, need to use multipart upload.
-    if (multipart_upload_id.empty() && last_part_size > upload_settings.max_single_part_upload_size)
-        createMultipartUpload();
+    detached_part_data.push_back(detachBuffer());
```

Comment:
The only thing you do with the result of `detachBuffer()` is appending it to the end of `detached_part_data`, so I think that appending can be made a part of `detachBuffer()`.

Reply:
Will do it